### ISSUE TITLE
Sidekiq hangs using Rails 6 app with "classic" autoloader

### ISSUE THREAD

**Issue by langsharpe** [rails/rails#36810]
Originally discussed on https://github.com/mperham/sidekiq/issues/4216 . This appears to be caused by a deadlock within actiontext.

### Steps to reproduce
- Create a new rails app using Rails 6.0.0.rc2
- Add Sidekiq and create some jobs (I used the instructions here https://github.com/mperham/sidekiq/wiki/Getting-Started#rails)
- Run `bundle exec sidekiq`
- Watch as it completes the jobs
- Switch to the classic autoloader (and do a `spring stop`)
- Create more jobs
- Run `bundle exec sidekiq`
- Watch as it starts jobs, but never completes any.

Here is a test app based on the instructions above: https://github.com/langsharpe/test_sidekiq_activetext

Here is a gist of the output of the process after sending the TTIN signal
https://gist.githubusercontent.com/langsharpe/8ca50bca283f863e09c226b0b88a3f57/raw/f8d6343fe2e388b6b2a16988dd281b29c57c332f/Sidekiq.log

Our use case is an existing app we upgraded to Rails 6, but is not Zeitwerk compatible (yet). The problem only occurs when code is autoloaded. It doesn't occur when code is eager loaded (e.g. in production mode)

### Expected behavior
Jobs should complete when using the classic autoloader.

### Actual behavior
Jobs start, but never finish.

### System configuration
**Rails version**: 6.0.0.rc2
**Ruby version**: 2.6.2p47 (2019-03-13 revision 67232) [x86_64-darwin18]

**Comment by terracatta** [comment]
@langsharpe I noticed this all the way back in the early betas. Couple of things that were relevant in our case:

1. The deadlocks only occurred in development, we never saw them in prod.
1. Running the sidekiq process with concurrency of `1` prevents the deadlocks

Do these match your observations as well?

**Comment by langsharpe** [comment]
> 1. The deadlocks only occurred in development, we never saw them in prod.
> 2. Running the sidekiq process with concurrency of `1` prevents the deadlocks
> 
> Do these match your observations as well?
Yes. Our workaround in development is to use a concurrency of 1.

**Comment by fxn** [comment]
Why do you start in `zeiwterk` mode and switch to `classic` in the middle of the reproduction? Can the instructions be simplified to starting in `classic` mode since the beginning?

**Comment by fxn** [comment]
Also, is it relevant to start with Spring enabled and stop it later? Can you reproduce disabling Spring from the start? (`export DISABLE_SPRING=1`).

**Comment by chrisyuska** [comment]
:+1: on this as I experienced the same issue with the same deadlock and following segmentation fault in ActionText from Sidekiq when concurrency > 1.  Two solutions fixed the problem for me:

1. Switch the autoloader to `:zeitwerk` (alternatively set `config.load_defaults '6.0'`)
2. Turn off ActionText via replacing `require 'rails/all'` in the application config with the contents of [railties/lib/rails/all.rb](https://github.com/rails/rails/blob/v6.0.0/railties/lib/rails/all.rb) with `action_text/engine` removed.

So for anyone else coming across this thread, there appears to be some deadlock issue with Rails's `classic` autoloader and ActionText, at least within Sidekiq.  Either of the solutions above should resolve the deadlock problem, but I haven't dug deeper into ActionText to investigate why.

**Comment by this-miguel** [comment]
I'm `config.autoloader = :classic` and `:concurrency: 1` in the `sidekiq.yml` seems to be working fine using sidekiq 6.0.0. Before I had `:concurrency: 5` and the workers were taking for ever to finish. The Sidekiq [update page to version 6](https://github.com/mperham/sidekiq/blob/master/6.0-Upgrade.md#welcome-to-sidekiq-60), state clearly that only `zeitwerk` mode works for rails 6.

**Comment by siebertm** [comment]
I managed to capture a backtrace of all the threads (using method described in )

```
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:114:in `sleep'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:114:in `wait'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:114:in `block (2 levels) in wait'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:113:in `handle_interrupt'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:113:in `block in wait'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:109:in `handle_interrupt'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:109:in `wait'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:128:in `wait_while'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/concurrency/share_lock.rb:221:in `wait_for'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/concurrency/share_lock.rb:83:in `block (2 levels) in start_exclusive'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/concurrency/share_lock.rb:82:in `block in start_exclusive'
~/.asdf/installs/ruby/2.6.5/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/concurrency/share_lock.rb:77:in `start_exclusive'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/concurrency/share_lock.rb:149:in `exclusive'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies/interlock.rb:13:in `loading'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:40:in `load_interlock'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:390:in `require_or_load'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:48:in `block in require_or_load'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:16:in `allow_bootsnap_retry'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:47:in `require_or_load'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:545:in `load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:60:in `block in load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:16:in `allow_bootsnap_retry'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:59:in `load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:214:in `const_missing'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:582:in `load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:60:in `block in load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:16:in `allow_bootsnap_retry'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:59:in `load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:214:in `const_missing'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:582:in `load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:60:in `block in load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:16:in `allow_bootsnap_retry'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:59:in `load_missing_constant'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:214:in `const_missing'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/actiontext-6.0.2.1/lib/action_text/engine.rb:46:in `block (2 levels) in <class:Engine>'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:72:in `class_eval'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:72:in `block in execute_hook'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:62:in `with_execution_control'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:67:in `execute_hook'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:52:in `block in run_load_hooks'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:51:in `each'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/lazy_load_hooks.rb:51:in `run_load_hooks'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/actiontext-6.0.2.1/lib/action_text/content.rb:132:in `<main>'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `block in require_with_bootsnap_lfi'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require_with_bootsnap_lfi'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:325:in `block in require'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:291:in `load_dependency'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/dependencies.rb:325:in `require'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/actiontext-6.0.2.1/lib/action_text/engine.rb:42:in `block (2 levels) in <class:Engine>'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:429:in `instance_exec'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:429:in `block in make_lambda'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:201:in `block (2 levels) in halting'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:607:in `block (2 levels) in default_terminator'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:606:in `catch'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:606:in `block in default_terminator'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:202:in `block in halting'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:514:in `block in invoke_before'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:514:in `each'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:514:in `invoke_before'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/callbacks.rb:134:in `run_callbacks'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/execution_wrapper.rb:111:in `run!'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/execution_wrapper.rb:73:in `block in run!'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/execution_wrapper.rb:70:in `tap'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/execution_wrapper.rb:70:in `run!'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/execution_wrapper.rb:86:in `wrap'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.2.1/lib/active_support/reloader.rb:71:in `wrap'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/rails.rb:42:in `call'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:132:in `block (4 levels) in dispatch'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:250:in `stats'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:127:in `block (3 levels) in dispatch'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/job_logger.rb:8:in `call'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:126:in `block (2 levels) in dispatch'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/job_retry.rb:74:in `global'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:125:in `block in dispatch'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/logging.rb:48:in `with_context'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/logging.rb:42:in `with_job_hash_context'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:124:in `dispatch'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:163:in `process'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:83:in `process_one'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/processor.rb:71:in `run'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/util.rb:16:in `watchdog'
~/.asdf/installs/ruby/2.6.5/lib/ruby/gems/2.6.0/gems/sidekiq-5.2.7/lib/sidekiq/util.rb:25:in `block in safe_thread'
```

Lots of noise here, but looking at `actiontext-6.0.2.1/lib/action_text/engine.rb:46` this might lead to the issue:


```ruby
    initializer "action_text.renderer" do |app|
      app.executor.to_run      { ActionText::Content.renderer = ApplicationController.renderer }
      app.executor.to_complete { ActionText::Content.renderer = ApplicationController.renderer }

      ActiveSupport.on_load(:action_text_content) do
        self.renderer = ApplicationController.renderer # <<= line 46
      end

      ActiveSupport.on_load(:action_controller_base) do
        before_action { ActionText::Content.renderer = ApplicationController.renderer.new(request.env) }
      end
    end
```

When I blindly comment out that line, it works.

**Comment by jmaino** [comment]
I also commented this in the sidekiq github but:

Hi everyone - I've had similar situations and found this thread.
I, like others potentially, cheated while upgrading and as-of 6.0 I began to see jobs stuck in busy as well. Moving to "load_defaults 6.0" seemed to work for me until I removed my old new_framework_defaults.rb (which I didn't think I needed anymore). Then the problem reappeared. Restored the file and traced to see what might have exposed this for me.

In my case I have it narrowed down to:
ActiveSupport.to_time_preserves_timezone = false

If the above is set, I see no hangs. If I comment this out or revert to the default true as-of 5.0, then I see the jobs stuck in the busy state again.

Perhaps I have something else wrongly or inconsistently configured - but wanted to share this in case it sheds any light on what might be the root cause and how to resolve it.

ruby 2.7.0
rails 6.0.2.2
sidekiq 6.0.6

**Comment by jgeiger** [comment]
Similar to @siebertm, I see the same line having an issue when I ctrl-c the running app and I get a core dump.

We're also running on an older rails app that has been upgraded from 3.2-> 5.x -> 6.x. We're still using the classic loader as well.

```
/gems/ruby/2.6.0/gems/actiontext-6.0.3.1/lib/action_text/engine.rb:46: [BUG] Segmentation fault at 0x0000561f14a9ccc0
ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0036 p:0009 s:0182 e:000179 BLOCK  /gems/ruby/2.6.0/gems/actiontext-6.0.3.1/lib/action_text/engine.rb:46 [FINISH]
c:0035 p:---- s:0177 e:000176 CFUNC  :instance_exec
c:0034 p:0036 s:0173 e:000172 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:428
c:0033 p:0011 s:0165 e:000164 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:200
c:0032 p:0006 s:0162 e:000161 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:605 [FINISH]
c:0031 p:---- s:0159 e:000158 CFUNC  :catch
c:0030 p:0012 s:0154 e:000153 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:604
c:0029 p:0045 s:0148 E:002398 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:201
c:0028 p:0008 s:0139 e:000138 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:513 [FINISH]
c:0027 p:---- s:0135 e:000134 CFUNC  :each
c:0026 p:0007 s:0131 e:000130 METHOD /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:513
c:0025 p:0085 s:0126 E:0012a8 METHOD /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:134
c:0024 p:0029 s:0117 e:000116 METHOD /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:111
c:0023 p:0009 s:0113 e:000112 BLOCK  /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:73 [FINISH]
c:0022 p:---- s:0108 e:000107 CFUNC  :tap
c:0021 p:0022 s:0104 e:000103 METHOD /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:70
c:0020 p:0013 s:0100 e:000099 METHOD /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:86
c:0019 p:0008 s:0095 e:000094 METHOD /gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/reloader.rb:71
c:0018 p:0010 s:0091 e:000090 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/rails.rb:42
c:0017 p:0008 s:0087 e:000086 BLOCK  /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:132
c:0016 p:0046 s:0084 e:000083 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:250
c:0015 p:0012 s:0077 e:000076 BLOCK  /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:127
c:0014 p:0002 s:0074 e:000073 METHOD /app/config/initializers/sidekiq.rb:32
c:0013 p:0014 s:0069 e:000068 BLOCK  /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:126
c:0012 p:0002 s:0066 e:000065 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/job_retry.rb:74
c:0011 p:0012 s:0057 e:000056 BLOCK  /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:125
c:0010 p:0054 s:0054 e:000053 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/logging.rb:48
c:0009 p:0014 s:0049 e:000048 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/logging.rb:42
c:0008 p:0023 s:0043 e:000042 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:124
c:0007 p:0047 s:0036 e:000035 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:163
c:0006 p:0019 s:0024 e:000023 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:83
c:0005 p:0010 s:0020 e:000019 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:71 [FINISH]
c:0004 p:---- s:0015 e:000014 IFUNC
c:0003 p:0002 s:0012 e:000011 METHOD /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/util.rb:16
c:0002 p:0029 s:0006 e:000005 BLOCK  /gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/util.rb:25 [FINISH]
c:0001 p:---- s:0003 e:000002 (none) [FINISH]

-- Ruby level backtrace information ----------------------------------------
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/util.rb:25:in `block in safe_thread'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/util.rb:16:in `watchdog'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:71:in `run'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:83:in `process_one'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:163:in `process'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:124:in `dispatch'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/logging.rb:42:in `with_job_hash_context'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/logging.rb:48:in `with_context'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:125:in `block in dispatch'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/job_retry.rb:74:in `global'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:126:in `block (2 levels) in dispatch'
/app/config/initializers/sidekiq.rb:32:in `call'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:127:in `block (3 levels) in dispatch'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:250:in `stats'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/processor.rb:132:in `block (4 levels) in dispatch'
/gems/ruby/2.6.0/gems/sidekiq-5.2.8/lib/sidekiq/rails.rb:42:in `call'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/reloader.rb:71:in `wrap'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:86:in `wrap'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:70:in `run!'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:70:in `tap'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:73:in `block in run!'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/execution_wrapper.rb:111:in `run!'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:134:in `run_callbacks'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:513:in `invoke_before'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:513:in `each'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:513:in `block in invoke_before'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:201:in `block in halting'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:604:in `block in default_terminator'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:604:in `catch'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:605:in `block (2 levels) in default_terminator'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:200:in `block (2 levels) in halting'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:428:in `block in make_lambda'
/gems/ruby/2.6.0/gems/activesupport-6.0.3.1/lib/active_support/callbacks.rb:428:in `instance_exec'
/gems/ruby/2.6.0/gems/actiontext-6.0.3.1/lib/action_text/engine.rb:46:in `block (2 levels) in <class:Engine>'
```

**Comment by p8** [comment]
@rafaelfranca I could reproduce this as fixed (probably by #40222)
The following sample app can reproduce it:
https://github.com/p8/rails-issue-36810

## Steps 

* Create a background job: 

   HardWorker.perform_async('bob', 5)

* Run `bundle exec sidekiq`

* Job runs on Rails master but does not run on Rails 6.0.3

cc @jonathanhefner

**Comment by ghiculescu** [comment]
@p8  if this is fixed, can we close the issue?

**Comment by p8** [comment]
@ghiculescu Still reproducible on 6.0.4, it's fixed on 6.1.0.

**Comment by p8** [comment]
@ghiculescu as this issue isn’t fixed in 6.0.4 and is on the 6.0.4 milestone, I think we should keep it open to allow upgrading to without skipping 6.0. I’ll try to find the commit that fixed it and backport it.

**Comment by p8** [comment]
This should be fixed by #42620

**Comment by jankeesvw** [comment]
I am seeing this bug, every time I make a change in my code Sidekiq freezes and won't proces any jobs until I restart the proces. Setting the concurrency to 1 seems to fix the problem.

**Comment by p8** [comment]
@jankeesvw
What version of Rails and Sidekiq are you using?

**Comment by jankeesvw** [comment]
@p8 I think I'm using the latest and greatest.

```bash
$ bundle info rails                                                                                                                                                                                                                         
# rails (7.0.3)
$ bundle info sidekiq                                                                                                                                                                                                                         
# sidekiq (6.5.1)
```

I'm using these sidekiq related gems:

```
gem "sidekiq", "~> 6.5"
gem "sidekiq-scheduler", "~> 4.0"
gem "sidekiq-unique-jobs", "~> 7.1.25"
```

**Comment by jankeesvw** [comment]
> I am seeing this bug, every time I make a change in my code Sidekiq freezes and won't proces any jobs until I restart the proces. Setting the concurrency to 1 seems to fix the problem.

Initially it looked like my problem was fixed, but unfortunately setting the concurrency to 1 did not fix it. The bug just takes a bit longer to show up.

**Comment by md-farhan-memon** [comment]
Was anybody able to fix this with `classic` autoloader? I am facing this issue too and it is on production.

Issue with log traces: mperham/sidekiq/issues/5453
Ruby version: 2.7.3
Rails version: 6.1.6
Sidekiq version(s): 6.5.1, 5.2.10, 4.2.10

**Comment by dcorking** [comment]
@md-farhan-memon @jankeesvw This is marked as fixed in Rails 6.1.0 (by https://github.com/rails/rails/commit/614e813161ad8f6e13ad19aedd577acdd976d9d6) and 6.0.4 (by https://github.com/rails/rails/pull/42620 ). Based on your descriptions, your bugs are not the same as this autoloading deadlock: you might have threading issues in other code. See https://github.com/sidekiq/sidekiq/wiki/Problems-and-Troubleshooting#threading

**Comment by md-farhan-memon** [comment]
We identified the issue and it was due to the [bug](https://bugs.ruby-lang.org/issues/17781) in ruby v2.7.3
It might help people coming here from future searches.
